src/Compilers/Core/Portable/SourceGeneration/Nodes/IncrementalValueSources.cs

```diff
@@ -13,21 +13,23 @@ namespace Microsoft.CodeAnalysis
     public readonly struct IncrementalValueSources
     {
         private readonly ArrayBuilder<ISyntaxInputNode> _syntaxInputBuilder;
+        private readonly ArrayBuilder<IIncrementalGeneratorOutputNode> _outputNodes;
 
-        internal IncrementalValueSources(ArrayBuilder<ISyntaxInputNode> syntaxInputBuilder)
+        internal IncrementalValueSources(ArrayBuilder<ISyntaxInputNode> syntaxInputBuilder, ArrayBuilder<IIncrementalGeneratorOutputNode> outputNodes)
         {
             _syntaxInputBuilder = syntaxInputBuilder;
+            _outputNodes = outputNodes;
         }
 
-        public SyntaxValueSources Syntax => new SyntaxValueSources(_syntaxInputBuilder);
+        public SyntaxValueSources Syntax => new SyntaxValueSources(_syntaxInputBuilder, _outputNodes.Add);
 
-        public IncrementalValueSource<Compilation> Compilation => new IncrementalValueSource<Compilation>(SharedInputNodes.Compilation);
+        public IncrementalValueSource<Compilation> Compilation => new IncrementalValueSource<Compilation>(SharedInputNodes.Compilation, _outputNodes.Add);
 
-        public IncrementalValueSource<ParseOptions> ParseOptions => new IncrementalValueSource<ParseOptions>(SharedInputNodes.ParseOptions);
+        public IncrementalValueSource<ParseOptions> ParseOptions => new IncrementalValueSource<ParseOptions>(SharedInputNodes.ParseOptions, _outputNodes.Add);
 
-        public IncrementalValueSource<AdditionalText> AdditionalTexts => new IncrementalValueSource<AdditionalText>(SharedInputNodes.AdditionalTexts);
+        public IncrementalValueSource<AdditionalText> AdditionalTexts => new IncrementalValueSource<AdditionalText>(SharedInputNodes.AdditionalTexts, _outputNodes.Add);
 
-        public IncrementalValueSource<AnalyzerConfigOptionsProvider> AnalyzerConfigOptions => new IncrementalValueSource<AnalyzerConfigOptionsProvider>(SharedInputNodes.AnalyzerConfigOptions);
+        public IncrementalValueSource<AnalyzerConfigOptionsProvider> AnalyzerConfigOptions => new IncrementalValueSource<AnalyzerConfigOptionsProvider>(SharedInputNodes.AnalyzerConfigOptions, _outputNodes.Add);
```

Comment:
Every call to `_outputNodes.Add` here is going to allocate a `delegate`. Consider pre-allocating that delegate in the `ctor`, storing it as a field and using that instead of `_outputNodes.Add`.